hw/xbox/mcpx_apu.c

```diff
@@ -826,12 +849,38 @@ static void se_frame(void *opaque)
                     NV_PAVS_VOICE_PAR_STATE_ACTIVE_VOICE)) {
                 MCPX_DPRINTF("voice %d not active...!\n", d->regs[current]);
                 fe_method(d, SE2FE_IDLE_VOICE, d->regs[current]);
+            } else {
+                process_voice(d, mixbins, d->regs[current]);
             }
             MCPX_DPRINTF("next voice %d\n", d->regs[next]);
             d->regs[current] = d->regs[next];
         }
     }
 
+#if GENERATE_MIXBIN_BEEP
+    /* Inject some audio to the mixbin for debugging.
+     * Signal is 1500 Hz sine wave, phase shifted by mixbin number. */
+    for (mixbin = 0; mixbin < NUM_MIXBINS; mixbin++) {
+        for (sample = 0; sample < NUM_SAMPLES_PER_FRAME; sample++) {
+            /* Avoid multiple of 1.0 / NUM_SAMPLES_PER_FRAME for phase shift,
+             * or waves cancel out */
+            float offset = mixbin / (float)NUM_SAMPLES_PER_FRAME -
+                           sample / (float)(NUM_SAMPLES_PER_FRAME + 1);
+            float wave = sinf(offset * M_PI * 2.0f);
+            mixbins[mixbin][sample] += wave * 0x3FFFFF;
+        }
+    }
+#endif
+
+    /* Write VP results to the GP DSP MIXBUF */
+    for (mixbin = 0; mixbin < NUM_MIXBINS; mixbin++) {
+        for (sample = 0; sample < NUM_SAMPLES_PER_FRAME; sample++) {
+            dsp_write_memory(d->gp.dsp, 'X', 0x001400 + mixbin * 0x20 + sample,
```

Comment:
Prefer `DSP_MIXBUFFER_BASE` instead of `0x001400` here

Reply:
I agree and will add it; I'll also change it for the existing occurrences.

[I originally didn't do it because we weren't sure about the naming / correctness](https://github.com/xqemu/xqemu/pull/92/files#r205284029) (we still aren't sure, but at least a bit more confident).

Ideally #158 (or something related) will solve all of these issues by providing a virtual memory space for MIXBUF. 
This would be similar to how it's done for MMIO, which has different sections for the standard DSP56k X, Y, P, but also the GP exclusive MIXBUF. The `char` based naming is horrible.

Reply:
I had to create a new macro in `mcpx_apu.c` as the `dsp/dsp_cpu.h` is for internal DSP use - the `dsp/dsp.h` interface is also too generic at the moment; so it made sense to move the GP specific code to `mcpx_apu.c`.

Eventually we'll have to rework the `dsp.h` interface to support different memory sizes and memory banks - also see my previous post and the linked issue.

I chose a slightly different macro name than the existing one, because we typically refer to MIXBUFFER as MIXBUF. The name is also getting long which is bad for our 80 column limit.
The one in the DSP code will probably be renamed or even removed in the future (depends on if we consider MIXBUF a different bank, or just part of X memory), because it seems, that the DSP shouldn't even have to know about the existence of the MIXBUF.